packetbeat/protos/cassandra/frame.go

```diff
@@ -0,0 +1,900 @@
+
+      0         8        16        24        32         40
+      +---------+---------+---------+---------+---------+
+      | version |  flags  |      stream       | opcode  |
+      +---------+---------+---------+---------+---------+
+      |                length                 |
+      +---------+---------+---------+---------+
+      |                                       |
+      .            ...  body ...              .
+      .                                       .
+      .                                       .
+      +----------------------------------------
+
+  The protocol is big-endian (network byte order).
+
+  some code derived from https://github.com/gocql/gocql
+
+*/
+package cassandra
+
+import (
+	"errors"
+	"fmt"
+	"github.com/elastic/beats/libbeat/logp"
+	"io"
+	"io/ioutil"
+	"net"
+	"runtime"
+	"sync"
+)
+
+var (
+	ErrFrameTooBig = errors.New("frame length is bigger than the maximum allowed")
+)
+
+type frameHeader struct {
+	version       protoVersion
+	flags         byte
+	stream        int
+	op            frameOp
+	length        int
+	customPayload map[string][]byte
+}
+
+func (f frameHeader) toMap() map[string]interface{} {
+	data := make(map[string]interface{})
+	data["version"] = fmt.Sprintf("%d", f.version.version())
+	data["flags"] = f.flags
+	data["stream"] = f.stream
+	data["op"] = f.op.String()
+	data["length"] = f.length
+	return data
+}
+
+func (f frameHeader) String() string {
+	return fmt.Sprintf("[header version=%s flags=0x%x stream=%d op=%s length=%d]", f.version.String(), f.flags, f.stream, f.op.String(), f.length)
+}
+
+func (f frameHeader) Header() frameHeader {
+	return f
+}
+
+const defaultBufSize = 128
+
+var framerPool = sync.Pool{
+	New: func() interface{} {
+		return &framer{
+			readBuffer: make([]byte, defaultBufSize),
+		}
+	},
+}
+
+// a framer is responsible for reading, writing and parsing frames on a single stream
+type framer struct {
+	r io.Reader
+
+	proto byte
+	// flags are for outgoing flags, enabling compression and tracing etc
+	flags    byte
+	compres  Compressor
+	headSize int
+	// if this frame was read then the header will be here
+	header *frameHeader
+
+	// if tracing flag is set this is not nil
+	traceID []byte
+
+	// holds a ref to the whole byte slice for rbuf so that it can be reset to
+	// 0 after a read.
+	readBuffer []byte
+
+	rbuf []byte
+}
+
+func newFramer(r io.Reader, compressor Compressor, version byte) *framer {
+	f := framerPool.Get().(*framer)
+	var flags byte
+	if compressor != nil {
+		flags |= flagCompress
+	}
+
+	version &= protoVersionMask
+
+	headSize := 8
+	if version > protoVersion2 {
+		headSize = 9
+	}
+
+	f.compres = compressor
+	f.proto = version
+	f.flags = flags
+	f.headSize = headSize
+
+	f.r = r
+	f.rbuf = f.readBuffer[:0]
+
+	f.header = nil
+	f.traceID = nil
+
+	return f
+}
+
+type frame interface {
+	Header() frameHeader
+}
+
+func readHeader(r io.Reader, p []byte) (head frameHeader, err error) {
```

Comment:
while readHeader can be rewritten to accept `*streambuf.Buffer` directly, maybe we shouldn't as this is directly copied from gocql (do not change original code too much).

Instead in parse check the header can be read `p.buf.Avail(9)` and pass the `*streambuf.Buffer` (which implements io.Reader) directly to readHeader. Store important header fields (or complete header if required) in parser + some state-flag checking the header must be parsed or not.